Update ReduceCceWorldtube with new features

knelli2 opened the pull request:
## Proposed changes

The worldtube data can now be metric modal/nodal or bondi modal/nodal and the executable can temporally combine H5 files before transforming the worldtube data into bondi modal form.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
There are two new options for the ReduceCceWorldtube executable. They are `InputDataFormat:` and `ExtractionRadius:`. Also the `InputH5File` option can now be either a single filename or a list of filenames to combine temporally.
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

~Depends on and includes #6396, #6397.~

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Update ReduceCceWorldube to take multiple input formats

The worldtube data can now be one of
MetricModal, MetricNodal, BondiModal, BondiNodal. Also, the executable
can now combine multiple H5 files into one.

knelli2 committed:
Rename ReduceCceWorldtube to PreprocessCceWorldtube

nilsdeppe commented on src/Executables/ReduceCceWorldtube/ReduceCceWorldtube.cpp:
```diff
@@ -224,27 +464,52 @@ struct LMaxFactor {
+      std::vector<std::string> operator()(const std::string& input) {
+        return {input};
+      }
+    };
```
FYI we do have a `Utilities/Overloader`, though here I'm not sure it matters which is used. Complexity is about the same.

nilsdeppe commented on src/Executables/ReduceCceWorldtube/ReduceCceWorldtube.cpp:
```diff
@@ -335,21 +632,106 @@ int main(int argc, char** argv) {
+                 : input_files[0];
+    };
+
+    const auto clean_temporary_file = [&]() {
```
`&temporary_combined_h5_file`. No need to capture everything.

nilsdeppe commented on src/Executables/ReduceCceWorldtube/ReduceCceWorldtube.cpp:
```diff
@@ -171,31 +217,213 @@ void perform_cce_worldtube_reduction(
+                          time_offset * number_of_angular_points,
+                      number_of_angular_points);
+  });
+  // NOLINTEND
```
I think these you can do with something like:
```cpp
make_const_view(&std::as_const(get(tuples::get<tag>(result)).data()),
                get(get<tag>(bondi_boundary_data)).data()),
                time_offset * number_of_angular_points,
                number_of_angular_points);
```
Then you don't need linting. Same below.

nilsdeppe commented on src/Executables/ReduceCceWorldtube/ReduceCceWorldtube.cpp:
```diff
@@ -171,31 +217,213 @@ void perform_cce_worldtube_reduction(
+      input_file, extraction_radius};
+  const size_t l_max = buffer_updater.get_l_max();
+
+  // we're not interpolating, this is just a reasonable number of rows to
```
interpolating in space or time? Could you clarify in the C++ comment please?

nilsdeppe commented on src/Executables/ReduceCceWorldtube/ReduceCceWorldtube.cpp:
```diff
@@ -171,31 +217,213 @@ void perform_cce_worldtube_reduction(
+  const size_t number_of_angular_points =
+      Spectral::Swsh::number_of_swsh_collocation_points(l_max);
+  const DataVector& time_buffer = buffer_updater.get_time_buffer();
+  // we're not interpolating, this is just a reasonable number of rows to
```
same, clarify space/time interpolation.

nilsdeppe commented on tests/Unit/Executables/CMakeLists.txt:
```diff
@@ -36,3 +36,33 @@ add_standalone_test(
   REGEX_TO_MATCH "AdamsBashforth\\[2\\][^\\n]*N *Y"
   )
+
+add_definitions(-DBINDIR="${CMAKE_BINARY_DIR}")
```
can you add this only to the specific target instead of all execs in the dir? I think `target_compile_definitions`

nilsdeppe commented on tests/Unit/Executables/Test_ReduceCceWorldtube.cpp:
```diff
@@ -0,0 +1,483 @@
+  check_expected_data(bondi_nodal_output_worldtube_filename, l_max,
+                      expected_data, second_expected_data, target_time,
+                      second_target_time, false);
+}
```
`return 0;`

nilsdeppe commented on tests/Unit/Executables/Test_ReduceCceWorldtube.cpp:
```diff
@@ -0,0 +1,483 @@
+#ifdef BINDIR
+  std::string executable = BINDIR;
+#else
+  std::string executable = "nothing";
```
should this just `ERROR`?

nilsdeppe commented on docs/Tutorials/CCE.md:
```diff
@@ -69,20 +69,104 @@ make CharacteristicExtract
+- Nodal cartesian metric data (which we refer to as "metric nodal")
+- Modal cartesian metric data (which we refer to as "metric modal")
+- Nodal Bondi-Sachs data (which we refer to as "bondi nodal")
+
+Requirements for these data formats are listed below.
+
```
We should probably state what we mean by `nodal` and `modal` because folks less familiar with spectral methods may not be familiar.

nilsdeppe commented on docs/Tutorials/CCE.md:
```diff
@@ -108,102 +194,50 @@ remaining values are the complex modes in m-varies-fastest format. That is,
+points are Gauss-Legendre in $cos(\theta)$ and equally spaced in $\phi$. Below
+is a routine for computing the spherical harmonic $\theta$ and $\phi$ values.
+These can be used to compute the Cartesian locations for a given radius using
+the standard transformation. The routine supports \f$\ell\in[4, 32]\f$.
```
Should we note (maybe even `\note`) that this is likely the easiest format for people to write out?

nilsdeppe commented on docs/Tutorials/CCE.md:
```diff
@@ -213,15 +247,29 @@ with a YAML file
+different checkpoints) and will combine the times from all H5 files alongside
+converting the worldtube data format. If there are duplicate or overlapping
+times, the last/latest of the times are chosen. If you pass multiple input
+worldtube H5 files, it is assumed that they are increasing in times.
```
`they are increasing in times.` -> `they are ordered increasing in time.`?

nilsdeppe commented on docs/Tutorials/CCE.md:
```diff
@@ -213,15 +247,29 @@ with a YAML file
+  `ExtractionRadius` option can be `Auto`. Otherwise, it must be specified.
+- The option `LMaxFactor` determines the factor by which the resolution of the
+  boundary computation that is run will exceed the resolution of the input and
+  output files. (Empirically, we have found that `LMaxFactor` of 3 is sufficient
+  to achieve roundoff precision in all boundary data we have attempted, and an
+  `LMaxFactor` of 2 is usually sufficient to vastly exceed the precision of the
+  simulation that provided the boundary dataset.)
```
I wouldn't put this in parens, just leave it here but no parens.

knelli2 committed:
Add static exec for writing worldtube coords to file

knelli2 commented:
@nilsdeppe I pushed a fixup and added a new commit that adds this small exec which writes the collocation points to a text file to the statically deployed execs. Then in the last commit I just additionally renamed the new things I added

nilsdeppe commented on docs/Tutorials/CCE.md:
```diff
@@ -199,6 +202,14 @@ supports \f$\ell\in[4, 32]\f$.
 \snippet Test_Spherepack.cpp spectre_cce_grid_point_locations
 </details>
 
+Alternatively, if your code can read in grid points from a text file, you can
```
text file or h5? Because the `-o` below is with a `.h5`. You can do this when you squash :)

nilsdeppe reviewed (commented):
LGTM! Please go ahead and squash, including the minor clarification request!
